uniffi_docs/src/lib.rs

```diff
@@ -0,0 +1,518 @@
+                Event::Text(s) => match current_stage {
+                    ParseStage::Description => {
+                        description_buff.push_str(&s);
+                        description_buff.push('\n');
+                    }
+                    ParseStage::Arguments => {
+                        if s.to_lowercase() == "arguments" {
+                            continue;
+                        }
+                        args_values_buff.push(s.to_string());
+                    }
+                    ParseStage::ReturnDescription => {
+                        if s.to_lowercase() == "returns" {
+                            continue;
+                        }
+                        return_description_buff.push_str(&s);
+                        return_description_buff.push('\n');
+                    }
+                },
+                Event::Code(s) => {
+                    args_keys_buff.push(s.to_string());
+                }
+                _ => (),
+            }
+        }
+
+        let mut arguments_descriptions = HashMap::with_capacity(args_keys_buff.len());
+        args_keys_buff
+            .into_iter()
+            .zip(args_values_buff.into_iter())
+            .for_each(|(k, v)| {
+                arguments_descriptions.insert(k, v.replace('-', "").trim().to_string());
+            });
+
+        let return_description = if return_description_buff.is_empty() {
+            None
+        } else {
+            Some(return_description_buff)
+        };
+
+        if arguments_descriptions.is_empty() && return_description.is_none() {
+            return Ok(Function {
+                description: s.to_string(),
+                arguments_descriptions,
+                return_description,
+            });
+        }
+
+        Ok(Function {
+            description: description_buff,
+            arguments_descriptions,
+            return_description,
+        })
+    }
+}
+
+/// Used to keep track of the different
+/// function comment parts while parsing it.
+enum ParseStage {
+    Description,
+    Arguments,
+    ReturnDescription,
+}
+
+/// Record or enum or object documentation.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct Structure {
+    pub description: String,
+
+    /// Members (record fields or enum variants) descriptions.
+    pub members: HashMap<String, String>,
+
+    /// Methods documentation - empty for records and enums.
+    pub methods: HashMap<String, Function>,
+}
+
+/// Impl documentation.
+#[derive(Debug, PartialEq, Eq)]
+struct Impl {
+    methods: HashMap<String, Function>,
+}
+
+#[derive(Debug, PartialEq, Eq)]
+pub struct Documentation {
+    pub functions: HashMap<String, Function>,
+    pub structures: HashMap<String, Structure>,
+}
+
+/// Extract doc comment from attributes.
+///
+/// Rust doc comments are silently converted (during parsing) to attributes of form:
+/// #[doc = "documentation comment content"]
+fn extract_doc_comment(attrs: &[Attribute]) -> Option<String> {
```

Comment:
How do you see this interacting with the proc-macros?  I don't love parsing the source code again when we already parse it in those macros.  Could this code be called directly from inside the macros?

Reply:
Yeah, for this reason I think your comment above about this being a separate crate is actually very much relevant. We should be doing this as part of the existing macros, not in an entirely different pass that does filesystem I/O.

Reply:
I'm guessing that since this review thread has not received new comments, nothing has changed in this regard?